### PR TITLE
support empty line comments

### PR DESCRIPTION
**Pull request by devoncarew**
- support empty line comments

This is to allow us to do things like:

```
// Copyright foo
// bar
// baz.

// Here's an additional file comment, but separated from the copyright header.
```

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

File: test/specs/library_test.dart [hunk 1]
```diff
@@ -45,7 +45,7 @@ void main() {
         ),
         equalsDart(r'''
           // Generated by foo!
-          //
```
Comment (srawlins) [inline test/specs/library_test.dart:48]: It seems like there should still be a way to get this output. Not sure how though... `comments` could be a list of nullable strings, and then null means something different from a blank string. Or a string that is only whitespace vs blank? I can't think of anything great...

Comment (devoncarew) [reply test/specs/library_test.dart:48]: I don't know that it's an important use case to support `//` lines? But the easiest and most clear way would likely be by supporting null strings.

An alternative approach would be to support line wrapping. So, multiple lines would mean multiple groups of comments, with each line automatically wrapped at ~80 cols.

I may implement the 1st solution as that seems like the smallest API diff.

**Comment by devoncarew** [comment]
ptal - updated to support both empty line comments as well as having groups of line comments

File: test/specs/library_test.dart (hunk 1 above)
Comment (srawlins) [reply test/specs/library_test.dart]: Well, I don't have any examples, but I can imagine that it's more common to write a multi-paragraph comment (with blank `//` lines) than adjacent comments. Like if you have two paragraphs, or an example, like a code block separated from test.

**Comment by srawlins** [comment]
I'm not a huge fan of the nullable String API, where null is sort of a "magic value." But I do think it is extremely pragmatic, functional, easy to understand...

I think I'd like another opinion like @natebosch before landing. If we go a nullable String API then it would be a pain to bring it back to a non-nullable String API. I'd almost prefer an API where "a comment" is a `List<String>` and a declaration, like a class, can have "plural comments", so a `List<Comment>`. But that might not be pragmatic, making everyone write so many nested structures just to add a little one-line comment somewhere...

**Comment by devoncarew** [comment]
> I'm not a huge fan of the nullable String API, where null is sort of a "magic value." But I do think it is extremely pragmatic, functional, easy to understand...

Agree w/ the nullability aspect; I think it doesn't improve the API. I've reverted back to a simplier form of this PR; happy to wait for an opinion from Nate.

For context, this PR is to improve the output of the files we generate for package:web.

**Comment by devoncarew** [comment]
Thinking about the comments above:

- I don't love changing the meaning of the current comments field (what this PR proposes and in Nate's idea above)
- I think a green field solution would use something like the `List<Comment>` idea from above, but I don't think the new api / deprecated api dance for a new `comments` field is worth it for this feature
- the use case I have is wanting both a file copyright section and a 'generated by' line for a generated file (and for that 2nd comment to be separate from the copyright one)

I have a separate PR now - https://github.com/dart-lang/code_builder/pull/441 - which adds a new field for that specific use case - a `generatedByComment` field. This is not dissimilar to the existing `Library.ignoreForFile` field.

**Pull request closed without merging**